### ISSUE TITLE
Lua: vim.api.nvim_set_hl() causes hang, massive memory leak

### ISSUE THREAD

**Issue by willeccles** [neovim/neovim#14753]
- `nvim --version`: `NVIM v0.5.0-dev+1283-g79a7b0c58-dirty`
- Operating system/version: Void Linux under WSL 1
- Terminal name/version: Windows Terminal (WSL 1) 1.8.1521.0
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

The space in the command is because I copy pasted from my Lua file which has a few commands neatly aligned.

```
nvim -u NORC
:lua vim.api.nvim_set_hl(0, 'User1',        {guibg='#3c3836', guifg='#928374'})
```

### Actual behaviour

Instantly hangs and uses >16GB of RAM within less than 30 seconds.

### Expected behaviour

Not entirely sure, as the behavior with 0 as the first argument is not clear, but ideally it would be the same as the corresponding `highlight User1 ...` command. Certainly not what happened, anyway.

This could potentially be a WSL-related issue, but it seems quite odd that a Lua function would cause this. Additionally, using `:lua vim.cmd[[hi User1...]]` works perfectly.

**Comment by shadmansaleh** [comment]
I thought I had seen this bug before. But can't reproduce it on master . Currently 1st argument 0 does nothing . It's planned to work like `:highlight` when ns is 0 but it's not implemented yet .

**Comment by bfredl** [comment]
Let's make it a proper error for 0.5

**Comment by lewis6991** [comment]
Should be fixed with #17187
